### PR TITLE
Patch insertAdjacentHTML

### LINKED ISSUE #104: Element#insertAdjacentHTML is unsupported
Custom elements inserted via `insertAdjacentHTML` are ignored / never upgraded.

<details>
<summary>Diff with failing tests</summary>

```diff
diff --git a/tests/html/Element/index.html b/tests/html/Element/index.html
index b771d37..872b872 100644
--- a/tests/html/Element/index.html
+++ b/tests/html/Element/index.html
@@ -9,5 +9,6 @@
     './removeAttribute.html',
     './removeAttributeNS.html',
     './insertAdjacentElement.html',
+    './insertAdjacentHTML.html'
   ]);
 </script>
diff --git a/tests/html/Element/insertAdjacentHTML.html b/tests/html/Element/insertAdjacentHTML.html
new file mode 100644
index 0000000..8dca1e6
--- /dev/null
+++ b/tests/html/Element/insertAdjacentHTML.html
@@ -0,0 +1,72 @@
+<!doctype html>
+<html>
+<head>
+<title>Element#insertAdjacentHTML</title>
+<script>
+  (window.customElements = window.customElements || {}).forcePolyfill = true;
+</script>
+<script src="../../../../es6-promise/dist/es6-promise.auto.min.js"></script>
+<script src="../../../../web-component-tester/browser.js"></script>
+<script src="../../../custom-elements.min.js"></script>
+</head>
+<body>
+<script>
+function generateLocalName() {
+  return 'test-element-' + Math.random().toString(32).substring(2);
+}
+
+function defineWithLocalName(localName, observedAttributes) {
+  customElements.define(localName, class extends HTMLElement {
+    constructor() {
+      super();
+      this.constructed = true;
+      this.connectedCallbackCount = 0;
+      this.disconnectedCallbackCount = 0;
+    }
+
+    connectedCallback() {
+      this.connectedCallbackCount++;
+    }
+
+    disconnectedCallback() {
+      this.disconnectedCallbackCount++;
+    }
+  });
+}
+
+suite('Custom elements in HTML strings are created using `insertAdjacentHTML`.', function() {
+  let localName;
+
+  setup(function() {
+    localName = generateLocalName();
+    defineWithLocalName(localName);
+  });
+
+  test('Disconnected context', function() {
+    const div = document.createElement('div');
+    div.insertAdjacentHTML('beforeend', `<${localName}></${localName}>`);
+
+    assert.equal(div.childNodes.length, 1);
+    assert.equal(div.childNodes[0].localName, localName);
+    assert.equal(div.childNodes[0].constructed, true);
+    assert.equal(div.childNodes[0].connectedCallbackCount, 0);
+    assert.equal(div.childNodes[0].disconnectedCallbackCount, 0);
+  });
+
+  test('Connected context', function() {
+    const div = document.createElement('div');
+    document.body.appendChild(div);
+    div.insertAdjacentHTML('beforeend', `<${localName}></${localName}>`);
+
+    assert.equal(div.childNodes.length, 1);
+    assert.equal(div.childNodes[0].localName, localName);
+    assert.equal(div.childNodes[0].constructed, true);
+    assert.equal(div.childNodes[0].connectedCallbackCount, 1);
+    assert.equal(div.childNodes[0].disconnectedCallbackCount, 0);
+
+    document.body.removeChild(div);
+  });
+});
+</script>
+</body>
+</html>
```
</details>

<details open>
<summary>Resulting failures</summary>
screen shot 2017-06-23 at 4 07 17 pm
</details>

### PR DESCRIPTION
**Pull request by muan**
Fixes #104.

Tests are _mostly_ passing locally. ⭐️  I am getting some flaky tests on Safari 11.0.2 on this branch as well as master so I'm disregarding them (`customElements.polyfillWrapFlushCallback is not a function`).

cc @bicknellr @notwaldorf ❤️
cc @josh @dgraham
cc @javan didn't find a commit for your test code, but would rebase on top of it if you have it somewhere 🙏

### CODE REVIEW

**Comment by javan** [comment]
> didn't find a commit for your test code

I only pasted a diff in https://github.com/webcomponents/custom-elements/issues/104. It's all yours. Thanks for fixing! 💝

File: src/Patch/Element.js [hunk 1]
```diff
@@ -228,6 +228,35 @@ export default function(internals) {
...
+  function patch_insertAdjacentHTML(destination, baseMethod) {
+    Utilities.setPropertyUnchecked(destination, 'insertAdjacentHTML',
+      /**
+       * @this {Element}
+       * @param {string} where
+       * @param {string} html
+       */
+      function(where, html) {
+        baseMethod.call(this, where, html);
+
```
Comment (bicknellr) [inline src/Patch/Element.js]: Could you add `where = where.toLowerCase();` here, before the comparison below? [This string doesn't have to match case-sensitively.](https://w3c.github.io/DOM-Parsing/#dom-element-insertadjacenthtml)

**Comment by bicknellr** [comment]
Also, you'll need to sign the CLA before I can pull this in and thanks for being consistent with the style!

**Comment by muan** [comment]
I signed it!

**Comment by bicknellr** [comment]
Sorry, I didn't mean that you needed overwrite all your stuff, just to pull in a copy of the tests!

**Comment by bicknellr** [comment]
Wait, I didn't read that commit correctly, live updates are funky. I'll just worry about coming up with the example of the thing I mentioned before and be back then.

**Comment by bicknellr** [comment]
Here's an example of the kind of situation I was referring to:
```javascript
test('cloneNode', function() {
  const log = [];

  class LoggingElement extends HTMLElement {
    constructor() {
      super();
      log.push(this.localName);
    }
  }

  const ceInsertedName = generateLocalName();
  const CEInserted = class extends LoggingElement {};
  customElements.define(ceInsertedName, CEInserted);

  const ce1Name = generateLocalName();
  customElements.define(ce1Name, class extends LoggingElement {
    constructor() {
      super();
      this.insertAdjacentHTML('afterend', `<${ceInsertedName}></${ceInsertedName}>`);
    }
  });

  const ce2Name = generateLocalName();
  customElements.define(ce2Name, class extends LoggingElement {});

  const template = document.createElement('template');
  template.innerHTML = `<div><${ce1Name}><${ce2Name}></${ce2Name}></${ce1Name}></div>`;

  // Insert a clone of the template, so that the elements are contained in a
  // document with a registry.

  const div = template.content.querySelector('div').cloneNode(true);
  document.body.appendChild(div);
  console.log('original:', div.outerHTML);
  console.log('log:', log);

  // Remove the <ce-inserted> created by the initial upgrade, so that the DOM
  // content looks like the HTML from the template.

  const inserted = div.querySelector(ceInsertedName);
  assert(inserted && inserted instanceof CEInserted);
  div.removeChild(inserted);
  console.log('original:', div.outerHTML);

  // Clear the upgrade log and clone the div.

  log.length = 0;
  const clone = div.cloneNode(true);

  console.log('clone:', clone.outerHTML);
  console.log('log:', log);

  // <ce-inserted> should have been upgraded before <ce-2>, even though it comes
  // after <ce-2> in document order.

  assert.deepEqual(log, [ce1Name, ceInsertedName, ce2Name]);
});
```

**Comment by muan** [comment]
@bicknellr I see what both of your comments mean now. 🙈 When I merged it in I thought I only see the tests. I didn't see there's a different insertAdjacentHTML patch.

Looks like your patch takes care of this test case 💯. Perhaps I just close this? :)

**Comment by bicknellr** [comment]
Oh, yeah it does seem like this branch includes both. Your call; if you want to keep going, sgtm, if not, I can pick up that other branch again.

**Pull request closed without merging**